README
NAME
    namespace::clean - Keep imports and functions out of your namespace

VERSION
    0.11

SYNOPSIS
      package Foo;
      use warnings;
      use strict;

      use Carp qw(croak);   # 'croak' will be removed

      sub bar { 23 }        # 'bar' will be removed

      # remove all previously defined functions
      use namespace::clean;

      sub baz { bar() }     # 'baz' still defined, 'bar' still bound

      # begin to collection function names from here again
      no namespace::clean;

      sub quux { baz() }    # 'quux' will be removed

      # remove all functions defined after the 'no' unimport
      use namespace::clean;

      # Will print: 'No', 'No', 'Yes' and 'No'
      print +(__PACKAGE__->can('croak') ? 'Yes' : 'No'), "\n";
      print +(__PACKAGE__->can('bar')   ? 'Yes' : 'No'), "\n";
      print +(__PACKAGE__->can('baz')   ? 'Yes' : 'No'), "\n";
      print +(__PACKAGE__->can('quux')  ? 'Yes' : 'No'), "\n";

      1;

DESCRIPTION
  Keeping packages clean
    When you define a function, or import one, into a Perl package, it will
    naturally also be available as a method. This does not per se cause
    problems, but it can complicate subclassing and, for example, plugin
    classes that are included via multiple inheritance by loading them as
    base classes.

    The "namespace::clean" pragma will remove all previously declared or
    imported symbols at the end of the current package's compile cycle.
    Functions called in the package itself will still be bound by their
    name, but they won't show up as methods on your class or instances.

    By unimporting via "no" you can tell "namespace::clean" to start
    collecting functions for the next "use namespace::clean;" specification.

    You can use the "-except" flag to tell "namespace::clean" that you don't
    want it to remove a certain function or method. A common use would be a
    module exporting an "import" method along with some functions:

      use ModuleExportingImport;
      use namespace::clean -except => [qw( import )];

    If you just want to "-except" a single sub, you can pass it directly.
    For more than one value you have to use an array reference.

  Explicitely removing functions when your scope is compiled
    It is also possible to explicitely tell "namespace::clean" what packages
    to remove when the surrounding scope has finished compiling. Here is an
    example:

      package Foo;
      use strict;

      # blessed NOT available

      sub my_class {
          use Scalar::Util qw( blessed );
          use namespace::clean qw( blessed );

          # blessed available
          return blessed shift;
      }

      # blessed NOT available

  Moose
    When using "namespace::clean" together with Moose you want to keep the
    installed "meta" method. So your classes should look like:

      package Foo;
      use Moose;
      use namespace::clean -except => 'meta';
      ...

    Same goes for Moose::Role.

  Cleaning other packages
    You can tell "namespace::clean" that you want to clean up another
    package instead of the one importing. To do this you have to pass in the
    "-cleanee" option like this:

      package My::MooseX::namespace::clean;
      use strict;

      use namespace::clean (); # no cleanup, just load

      sub import {
          namespace::clean->import(
            -cleanee => scalar(caller),
            -except  => 'meta',
          );
      }

    If you don't care about "namespace::clean"s discover-and-"-except"
    logic, and just want to remove subroutines, try "clean_subroutines".

METHODS
    You shouldn't need to call any of these. Just "use" the package at the
    appropriate place.

  clean_subroutines
    This exposes the actual subroutine-removal logic.

      namespace::clean->clean_subroutines($cleanee, qw( subA subB ));

    will remove "subA" and "subB" from $cleanee. Note that this will remove
    the subroutines immediately and not wait for scope end. If you want to
    have this effect at a specific time (e.g. "namespace::clean" acts on
    scope compile end) it is your responsibility to make sure it runs at
    that time.

  import
    Makes a snapshot of the current defined functions and installs a
    B::Hooks::EndOfScope hook in the current scope to invoke the cleanups.

  unimport
    This method will be called when you do a

      no namespace::clean;

    It will start a new section of code that defines functions to clean up.

  get_class_store
    This returns a reference to a hash in a passed package containing
    information about function names included and excluded from removal.

  get_functions
    Takes a class as argument and returns all currently defined functions in
    it as a hash reference with the function name as key and a typeglob
    reference to the symbol as value.

IMPLEMENTATION DETAILS
    This module works through the effect that a

      delete $SomePackage::{foo};

    will remove the "foo" symbol from $SomePackage for run time lookups
    (e.g., method calls) but will leave the entry alive to be called by
    already resolved names in the package itself. "namespace::clean" will
    restore and therefor in effect keep all glob slots that aren't "CODE".

    A test file has been added to the perl core to ensure that this
    behaviour will be stable in future releases.

    Just for completeness sake, if you want to remove the symbol completely,
    use "undef" instead.

SEE ALSO
    B::Hooks::EndOfScope

AUTHOR AND COPYRIGHT
    Robert 'phaylon' Sedlacek "<rs@474.at>", with many thanks to Matt S
    Trout for the inspiration on the whole idea.

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the same terms as perl itself.



Changes
    [0.20]
        - Bump Package::Stash dependency to 0.22 to pull in a bugfix in
          Package::Stash::XS 0.19.

    [0.19]
        - Port to the new Package::Stash 0.18 API and depend on it.
        - Don't rely on package::stash's remove_package_symbol implementation
          (doy).

    [0.18]
        - Make sure we continue working on future Package::Stash
          versions (doy).

    [0.17]
        - Make sure the debugger author test is skipped for
          non-authors before the debugger is even loaded.

    [0.16]
        - Release all changes of 0.15 as part of a stable release.
        - Convert from Module::Install to Dist::Zilla.

    [0.15] TRIAL release
        - Use Package::Stash for the stash manipulation bits (doy).

    [0.14] Thu Mar 18 11:15:38 CET 2010
        - Disable auto_install.
        - Turn the error prone debugger test into an author test.

    [0.13] Sun Jan 17 02:40:48 CET 2010
        - Skip failing debugger tests on 5.8.8 and older.

    [0.12] Thu Jan 14 03:22:03 CET 2010
        - Stop relying on stash entries always being upgraded into real GVs
          (Zefram).
        - Work around $DB::sub (Yuval Kogman).
        - Fix restoring of non-code symbols when cleaning (Ben Morrows).

    [0.11] Tue Mar  3 17:34:49 CET 2009
        - Added -cleanee option to specify the package to clean (Closes
          RT#41850).
        - Added n:c->clean_subroutines($cleanee, @subs).

    [0.10] Fri Feb 20 14:31:36 CET 2009
        - Depend on B::Hooks::EndOfScope 0.07 to avoid segfaults and lost error
          messages when something goes wrong.

    [0.09] Wed Oct 22 17:48:49 CEST 2008
        - Use B::Hooks::EndOfScope instead of %^H + Scope::Guard.

    [0.08] Sun Mar  9 22:01:01 CET 2008
        - Added explicit cleanup behaviour

    [0.07] Sun Mar  9 20:13:33 CET 2008
        - Switched from Filter::EOF to a much saner implementation via
          %^H and Scope::Guard. (mst & autobox)++ for this.

    [0.06] Wed Feb 20 15:09:00 CET 2008
        - Fixed 'uninitialized value in ref-to-glob cast' error
          if unimport was used before.

    [0.05] Sun Aug 12 18:24:49 CEST 2007
        - Minor POD improvements
        - -except now accepts a single value too

    [0.04] Sat Mar 17 16:22:10 CET 2007
        - Added -except flag
        - Non-CODE type slots will not be removed

    [0.03] Sat Feb 24 22:34:55 CET 2007
        - Minor comment and POD cleanups
        - Tried to clarify how the module works

    [0.02] Tue Feb 20 00:38:24 CET 2007
        - Added unimport behaviour

    [0.01] Sun Feb 18 17:33:18 CET 2007
        - Initial Version



t/lib/ExporterTest.pm
package ExporterTest;
use warnings;
use strict;

use base 'Exporter';
use vars qw( @EXPORT_OK );

@EXPORT_OK = qw( foo bar );

sub foo { 23 }
sub bar { 12 }

1;


t/lib/FunctionWipeout.pm
package FunctionWipeout;
use warnings;
use strict;

use ExporterTest qw(foo);

sub bar { foo() }

use namespace::clean;

sub baz { bar() }

1;


Makefile.PL
#!/usr/bin/env perl
use warnings;
use strict;

use inc::Module::Install;

name            q{namespace-clean};
license         q{perl};
author          q{Robert 'phaylon' Sedlacek <rs@474.at>};
all_from        q{lib/namespace/clean.pm};

build_requires  q{Test::More},                  '0.88';
build_requires  q{FindBin},                     0;

requires        q{B::Hooks::EndOfScope},        '0.07';
requires        q{Package::Stash},              '0.03';
requires        q{Sub::Name},                   '0.04';
requires        q{Sub::Identify},               '0.04';

auto_provides;

if (-e 'MANIFEST.SKIP') {

    print "Creating README file\n";
    system 'pod2text lib/namespace/clean.pm > README';

    print "Adding author tests to 'make test' run\n";
    tests 't/*.t t_author/*.t';
}

WriteAll;


lib/namespace/clean.pm
package namespace::clean;
# ABSTRACT: Keep imports and functions out of your namespace

use warnings;
use strict;

use vars qw( $STORAGE_VAR );
use Sub::Name 0.04 qw(subname);
use Sub::Identify 0.04 qw(sub_fullname);
use Package::Stash 0.22;
use B::Hooks::EndOfScope 0.07;

$STORAGE_VAR = '__NAMESPACE_CLEAN_STORAGE';

=head1 SYNOPSIS

  package Foo;
  use warnings;
  use strict;

  use Carp qw(croak);   # 'croak' will be removed

  sub bar { 23 }        # 'bar' will be removed

  # remove all previously defined functions
  use namespace::clean;

  sub baz { bar() }     # 'baz' still defined, 'bar' still bound

  # begin to collection function names from here again
  no namespace::clean;

  sub quux { baz() }    # 'quux' will be removed

  # remove all functions defined after the 'no' unimport
  use namespace::clean;

  # Will print: 'No', 'No', 'Yes' and 'No'
  print +(__PACKAGE__->can('croak') ? 'Yes' : 'No'), "\n";
  print +(__PACKAGE__->can('bar')   ? 'Yes' : 'No'), "\n";
  print +(__PACKAGE__->can('baz')   ? 'Yes' : 'No'), "\n";
  print +(__PACKAGE__->can('quux')  ? 'Yes' : 'No'), "\n";

  1;

=head1 DESCRIPTION

=head2 Keeping packages clean

When you define a function, or import one, into a Perl package, it will
naturally also be available as a method. This does not per se cause
problems, but it can complicate subclassing and, for example, plugin
classes that are included via multiple inheritance by loading them as
base classes.

The C<namespace::clean> pragma will remove all previously declared or
imported symbols at the end of the current package's compile cycle.
Functions called in the package itself will still be bound by their
name, but they won't show up as methods on your class or instances.

By unimporting via C<no> you can tell C<namespace::clean> to start
collecting functions for the next C<use namespace::clean;> specification.

You can use the C<-except> flag to tell C<namespace::clean> that you
don't want it to remove a certain function or method. A common use would
be a module exporting an C<import> method along with some functions:

  use ModuleExportingImport;
  use namespace::clean -except => [qw( import )];

If you just want to C<-except> a single sub, you can pass it directly.
For more than one value you have to use an array reference.

=head2 Explicitely removing functions when your scope is compiled

It is also possible to explicitely tell C<namespace::clean> what packages
to remove when the surrounding scope has finished compiling. Here is an
example:

  package Foo;
  use strict;

  # blessed NOT available

  sub my_class {
      use Scalar::Util qw( blessed );
      use namespace::clean qw( blessed );

      # blessed available
      return blessed shift;
  }

  # blessed NOT available

=head2 Moose

When using C<namespace::clean> together with L<Moose> you want to keep
the installed C<meta> method. So your classes should look like:

  package Foo;
  use Moose;
  use namespace::clean -except => 'meta';
  ...

Same goes for L<Moose::Role>.

=head2 Cleaning other packages

You can tell C<namespace::clean> that you want to clean up another package
instead of the one importing. To do this you have to pass in the C<-cleanee>
option like this:

  package My::MooseX::namespace::clean;
  use strict;

  use namespace::clean (); # no cleanup, just load

  sub import {
      namespace::clean->import(
        -cleanee => scalar(caller),
        -except  => 'meta',
      );
  }

If you don't care about C<namespace::clean>s discover-and-C<-except> logic, and
just want to remove subroutines, try L</clean_subroutines>.

=method clean_subroutines

This exposes the actual subroutine-removal logic.

  namespace::clean->clean_subroutines($cleanee, qw( subA subB ));

will remove C<subA> and C<subB> from C<$cleanee>. Note that this will remove the
subroutines B<immediately> and not wait for scope end. If you want to have this
effect at a specific time (e.g. C<namespace::clean> acts on scope compile end)
it is your responsibility to make sure it runs at that time.

=cut

my $RemoveSubs = sub {

    my $cleanee = shift;
    my $store   = shift;
    my $cleanee_stash = Package::Stash->new($cleanee);
    my $deleted_stash = Package::Stash->new("namespace::clean::deleted::$cleanee");
  SYMBOL:
    for my $f (@_) {
        my $variable = "&$f";
        # ignore already removed symbols
        next SYMBOL if $store->{exclude}{ $f };

        next SYMBOL unless $cleanee_stash->has_symbol($variable);

        if (ref(\$cleanee_stash->namespace->{$f}) eq 'GLOB') {
            # convince the Perl debugger to work
            # it assumes that sub_fullname($sub) can always be used to find the CV again
            # since we are deleting the glob where the subroutine was originally
            # defined, that assumption no longer holds, so we need to move it
            # elsewhere and point the CV's name to the new glob.
            my $sub = $cleanee_stash->get_symbol($variable);
            if ( sub_fullname($sub) eq ($cleanee_stash->name . "::$f") ) {
                my $new_fq = $deleted_stash->name . "::$f";
                subname($new_fq, $sub);
                $deleted_stash->add_symbol($variable, $sub);
            }
        }

        my ($scalar, $array, $hash, $io) = map {
            $cleanee_stash->get_symbol($_ . $f)
        } '$', '@', '%', '';
        $cleanee_stash->remove_glob($f);
        for my $var (['$', $scalar], ['@', $array], ['%', $hash], ['', $io]) {
            next unless defined $var->[1];
            $cleanee_stash->add_symbol($var->[0] . $f, $var->[1]);
        }
    }
};

sub clean_subroutines {
    my ($nc, $cleanee, @subs) = @_;
    $RemoveSubs->($cleanee, {}, @subs);
}

=method import

Makes a snapshot of the current defined functions and installs a
L<B::Hooks::EndOfScope> hook in the current scope to invoke the cleanups.

=cut

sub import {
    my ($pragma, @args) = @_;

    my (%args, $is_explicit);

  ARG:
    while (@args) {

        if ($args[0] =~ /^\-/) {
            my $key = shift @args;
            my $value = shift @args;
            $args{ $key } = $value;
        }
        else {
            $is_explicit++;
            last ARG;
        }
    }

    my $cleanee = exists $args{ -cleanee } ? $args{ -cleanee } : scalar caller;
    if ($is_explicit) {
        on_scope_end {
            $RemoveSubs->($cleanee, {}, @args);
        };
    }
    else {

        # calling class, all current functions and our storage
        my $functions = $pragma->get_functions($cleanee);
        my $store     = $pragma->get_class_store($cleanee);
        my $stash     = Package::Stash->new($cleanee);

        # except parameter can be array ref or single value
        my %except = map {( $_ => 1 )} (
            $args{ -except }
            ? ( ref $args{ -except } eq 'ARRAY' ? @{ $args{ -except } } : $args{ -except } )
            : ()
        );

        # register symbols for removal, if they have a CODE entry
        for my $f (keys %$functions) {
            next if     $except{ $f };
            next unless $stash->has_symbol("&$f");
            $store->{remove}{ $f } = 1;
        }

        # register EOF handler on first call to import
        unless ($store->{handler_is_installed}) {
            on_scope_end {
                $RemoveSubs->($cleanee, $store, keys %{ $store->{remove} });
            };
            $store->{handler_is_installed} = 1;
        }

        return 1;
    }
}

=method unimport

This method will be called when you do a

  no namespace::clean;

It will start a new section of code that defines functions to clean up.

=cut

sub unimport {
    my ($pragma, %args) = @_;

    # the calling class, the current functions and our storage
    my $cleanee   = exists $args{ -cleanee } ? $args{ -cleanee } : scalar caller;
    my $functions = $pragma->get_functions($cleanee);
    my $store     = $pragma->get_class_store($cleanee);

    # register all unknown previous functions as excluded
    for my $f (keys %$functions) {
        next if $store->{remove}{ $f }
             or $store->{exclude}{ $f };
        $store->{exclude}{ $f } = 1;
    }

    return 1;
}

=method get_class_store

This returns a reference to a hash in a passed package containing
information about function names included and excluded from removal.

=cut

sub get_class_store {
    my ($pragma, $class) = @_;
    my $stash = Package::Stash->new($class);
    my $var = "%$STORAGE_VAR";
    $stash->add_symbol($var, {})
        unless $stash->has_symbol($var);
    return $stash->get_symbol($var);
}

=method get_functions

Takes a class as argument and returns all currently defined functions
in it as a hash reference with the function name as key and a typeglob
reference to the symbol as value.

=cut

sub get_functions {
    my ($pragma, $class) = @_;

    my $stash = Package::Stash->new($class);
    return {
        map { $_ => $stash->get_symbol("&$_") }
            $stash->list_all_symbols('CODE')
    };
}

=head1 IMPLEMENTATION DETAILS

This module works through the effect that a

  delete $SomePackage::{foo};

will remove the C<foo> symbol from C<$SomePackage> for run time lookups
(e.g., method calls) but will leave the entry alive to be called by
already resolved names in the package itself. C<namespace::clean> will
restore and therefor in effect keep all glob slots that aren't C<CODE>.

A test file has been added to the perl core to ensure that this behaviour
will be stable in future releases.

Just for completeness sake, if you want to remove the symbol completely,
use C<undef> instead.

=head1 SEE ALSO

L<B::Hooks::EndOfScope>

=head1 THANKS

Many thanks to Matt S Trout for the inspiration on the whole idea.

=cut

no warnings;
'Danger! Laws of Thermodynamics may not apply.'
